<div class="cbp-l-inline">
    <div class="cbp-l-inline-left">
        <div class="cbp-slider">
            <ul class="cbp-slider-wrap">
                <li class="cbp-slider-item">
                    <a href="assets/pages/img/home/img1.png" class="cbp-lightbox">
                        <img src="assets/pages/img/home/img1.png" alt=""> </a>
                </li>
                <li class="cbp-slider-item">
                    <a href="assets/pages/img/home/img2.png" class="cbp-lightbox">
                        <img src="assets/pages/img/home/img2.png" alt=""> </a>
                </li>
             </ul>
        </div>
    </div>
    <div class="cbp-l-inline-right">
        <div class="cbp-l-inline-title">TADbit</div>
        <div class="cbp-l-inline-subtitle">by Fran√ßois Serra</div>
        <div class="cbp-l-inline-desc">
					<p>TADbit is a complete Python library to deal with all steps to analyze, model and explore 3C-based data. With TADbit the user can map FASTQ files to obtain raw interaction binned matrices (Hi-C like matrices), normalize and correct interaction matrices, identify adn compare the so-called Topologically Associating Domains (TADs), build 3D models from the interaction matrices, and finally, extract structural properties from the models. TADbit is complemented by TADkit for visualizing 3D models.</p>

					<p>Hi-C experiments generate genomic interaction between loci located in the same or in different chromosomes. TADbit is built around the concept of a chromosome, and uses it as a central item to store and compare different Hi-C experiments. The library has been designed to be used by researchers with no expertise in computer science. All-in-one scripts provided in TADbit allow to run the full analysis using one single command line; advanced users may produce their own programs using TADbit as a complementary library.</p>
				</div>
        <a href="workspace/"  class="cbp-l-inline-view btn blue btn-outline">RUN TOOL</a>
    </div>
</div>
